### PR TITLE
OnkyoAVTCP: Add source selection, detection via NRIQSTN

### PR DESCRIPTION
**Pull request by dwmw2**
Now it detects what the names of all the sources are, and lets me choose them. Also has support for 4 zones and automatically enables the ones that exist in the detected hardware.

Obsoletes #1958 and #1960

### CODE REVIEW

**Comment by dwmw2** [comment]
```
..\hardware\OnkyoAVTCP.cpp(192): error C3480: 'ID_MVL': a lambda capture variable must be from an enclosing function scope [C:\projects\domoticz\msbuild\domoticz.vcxproj]
..\hardware\OnkyoAVTCP.cpp(192): error C3260: '=': skipping unexpected token(s) before lambda body [C:\projects\domoticz\msbuild\domoticz.vcxproj]
..\hardware\OnkyoAVTCP.cpp(192): error C2143: syntax error: missing ';' before '}' [C:\projects\domoticz\msbuild\domoticz.vcxproj]
```

FFS. Is that just because Microsoft haven't yet managed to implement a C standard from the last century? What is the MSVC-safe way to do explicit initialisers? This is the offending code:
```
static struct {
	…
} switch_types[] = {
	[ID_MVL] = { "MVL", "AMT", "Master volume", STYPE_Dimmer, sSwitchGeneralSwitch, 8, NULL },
	[ID_ZVL] = { "ZVL", "ZMT", "Zone 2 volume", STYPE_Dimmer, sSwitchGeneralSwitch, 8, NULL },
	…
};
```

**Comment by dwmw2** [comment]
I killed the explicit initialisation. It was all in order with no gaps anyway; I just wanted to be extra safe.

File: hardware/OnkyoAVTCP.cpp [hunk 1]
```diff
@@ -305,26 +356,157 @@ void OnkyoAVTCP::ReceiveSwitchMsg(const char *pData, int Len, bool muting, int I
...
 
-	if (result.empty()) {
-		m_mainworker.PushAndWaitRxMessage(this, (const unsigned char *)&gswitch, switch_types[ID].name, 255);
-		m_sql.safe_query("UPDATE DeviceStatus SET SwitchType=%d, CustomImage=%d WHERE(HardwareID == %d) AND (DeviceID == '%08x')", switch_types[ID].switchType, switch_types[ID].customImage, m_HwdID, ID);
-		if (switch_types[ID].options) {
-			result = m_sql.safe_query("SELECT ID FROM DeviceStatus WHERE (HardwareID==%d) AND (DeviceID=='%08x') AND (Type==%d) ", m_HwdID, ID, pTypeGeneralSwitch);
-			if (result.size() > 0) {
-				std::string Idx = result[0][0];
-				m_sql.SetDeviceOptions(atoi(Idx.c_str()), m_sql.BuildDeviceOptions(switch_types[ID].options, false));
+	sDecodeRXMessage(this, (const unsigned char *)&gswitch, switch_types[ID].name, 255);
```
Comment (gizmocuz) [inline hardware/OnkyoAVTCP.cpp:359]: Please use the default SendSwitch function

File: hardware/OnkyoAVTCP.cpp [hunk 2]
```diff
@@ -364,7 +546,7 @@ void OnkyoAVTCP::ParseData(const unsigned char *pData, int Len)
 			continue;
 		}
 		int data_size = static_cast<int>(ntohl(pkt->data_size));
-//		_log.Log(LOG_NORM, "ISCP at len %d size %d", Len, data_size);
+		_log.Log(LOG_NORM, "ISCP at len %d size %d", Len, data_size);
```
Comment (gizmocuz) [inline hardware/OnkyoAVTCP.cpp]: I think this is a debug log and should be removed?

File: hardware/OnkyoAVTCP.h [hunk 3]
```diff
@@ -3,6 +3,7 @@
 #include <iosfwd>
 #include "ASyncTCP.h"
 #include "DomoticzHardware.h"
+#include <map>
```
Comment (gizmocuz) [inline hardware/OnkyoAVTCP.h]: I see here a map include but no reference to a map usage in this header? If it is used in the .cpp, then please include this in the .cpp only

File: main/RFXNames.cpp [hunk 4]
```diff
@@ -1963,7 +1963,7 @@ void GetSelectorSwitchStatuses(const std::map<std::string, std::string> & option
 		std::vector<std::string>::iterator itt;
 		int i = 0;
 		std::stringstream ss;
-		for (itt = strarray.begin(); (itt != strarray.end()) && (i <= 100); ++itt) {
+		for (itt = strarray.begin(); (itt != strarray.end()); ++itt) {
```
Comment (gizmocuz) [inline main/RFXNames.cpp:1966]: In the feature please make two separate pull requests because this one targets two different issues.
If for any reason we need to revert this pull request, this patch will also be removed, or visa versa

File: main/SQLHelper.cpp [hunk 5]
```diff
@@ -7820,14 +7815,25 @@ bool CSQLHelper::SetDeviceOptions(const uint64_t idx, const std::map<std::string
...
+}
+
+bool CSQLHelper::SetDeviceOptions(const uint64_t idx, const std::map<std::string, std::string> & optionsMap) {
+	if (idx < 1) {
+		_log.Log(LOG_ERROR, "Cannot set options on device %" PRIu64 "", idx);
 		return false;
 	}
-	if (options.empty()) {
+
+	if (optionsMap.size() == 0) {
```
Comment (gizmocuz) [inline main/SQLHelper.cpp]: this should be optionsMap.empty() please

File: main/SQLHelper.cpp [hunk 6]
```diff
@@ -7795,12 +7795,7 @@ std::map<std::string, std::string> CSQLHelper::GetDeviceOptions(const std::strin
 	return optionsMap;
 }
 
-bool CSQLHelper::SetDeviceOptions(const uint64_t idx, const std::map<std::string, std::string> & optionsMap) {
-	if (idx < 1) {
-		_log.Log(LOG_ERROR, "Cannot set options on device %" PRIu64 "", idx);
-		return false;
-	}
-
+std::string CSQLHelper::FormatDeviceOptions(const std::map<std::string, std::string> & optionsMap) {
```
Comment (gizmocuz) [inline main/SQLHelper.cpp:7798]: optionsMap could be declared const

File: main/SQLHelper.cpp [hunk 7]
```diff
@@ -7820,14 +7815,25 @@ bool CSQLHelper::SetDeviceOptions(const uint64_t idx, const std::map<std::string
 		}
 		options.assign(ssoptions.str());
 	}
-	if (options.empty() && (count > 0)) {
-		_log.Log(LOG_ERROR, "Cannot parse options for device %" PRIu64 "", idx);
+
+	return options;
+}
+
+bool CSQLHelper::SetDeviceOptions(const uint64_t idx, const std::map<std::string, std::string> & optionsMap) {
```
Comment (gizmocuz) [inline main/SQLHelper.cpp]: optionsMap could be declared const

File: main/SQLHelper.h [hunk 8]
```diff
@@ -361,6 +361,7 @@ class CSQLHelper
 
 	std::map<std::string, std::string> BuildDeviceOptions(const std::string & options, const bool decode = true);
 	std::map<std::string, std::string> GetDeviceOptions(const std::string & idx);
+	std::string FormatDeviceOptions(const std::map<std::string, std::string> & optionsMap);
```
Comment (gizmocuz) [inline main/SQLHelper.h]: optionsMap could be declared const

File: hardware/OnkyoAVTCP.cpp (hunk 1 above)
Comment (dwmw2) [reply hardware/OnkyoAVTCP.cpp:359]: Ah yes, I think I can so that now that I've made the device creation saner. Will fix.

File: hardware/OnkyoAVTCP.h (hunk 3 above)
Comment (dwmw2) [reply hardware/OnkyoAVTCP.h]: Yep, there *was* a map but I simplified things. Will fix it.

File: main/RFXNames.cpp (hunk 4 above)
Comment (dwmw2) [reply main/RFXNames.cpp:1966]: I can do that right now if you prefer. The disadvantage is that then we have two separate pull requests that depend on each other and if you revert the other one, you break this one anyway.

Comment (gizmocuz) [reply main/RFXNames.cpp:1966]: Plz leave it in.. i've seen the other pull request, and both look fine!

File: hardware/OnkyoAVTCP.cpp (hunk 1 above)
Comment (dwmw2) [reply hardware/OnkyoAVTCP.cpp:359]: Hm, maybe not... that sends a lighting2 message, and this is a generalswitch. I also do still have special magic to do, to look up the target level for the selector from LevelActions. I'm not quite sure I see how to make it use SendSwitch, or at least I don't see that it simplifies anything. Am I missing something?

Comment (gizmocuz) [reply hardware/OnkyoAVTCP.cpp:359]: Does it matter is it is a lighting2 or general switch ? Or are you using a selector switch ?
There might be a SendSelectorSwitch option soon, till then please use you current method

File: main/SQLHelper.cpp (hunk 6 above)
Comment (dwmw2) [reply main/SQLHelper.cpp:7798]: Um... isn't it?
```const map<string, string> & optionsMap```

File: main/SQLHelper.cpp (hunk 5 above)
Comment (dwmw2) [reply main/SQLHelper.cpp]: Fixed.

File: hardware/OnkyoAVTCP.cpp (hunk 2 above)
Comment (dwmw2) [reply hardware/OnkyoAVTCP.cpp]: Removed.

File: main/SQLHelper.cpp (hunk 6 above)
Comment (gizmocuz) [reply main/SQLHelper.cpp:7798]: Whoops sorry..

**Comment by dwmw2** [comment]
I've addressed the feedback I could understand.
Not done: 
 • make optionsMap more const
 • use SendSwitch()

As far as I can tell, using SendSwitch() wouild require changing the type of the underlying devices to a lighting switch?

File: hardware/OnkyoAVTCP.cpp (hunk 1 above)
Comment (dwmw2) [reply hardware/OnkyoAVTCP.cpp:359]: I have selector, on/off and level switches. Currently they are all `pTypeGeneralSwitch` devices and handled generically, based on the following table:

```
	{ "MVL", "AMT", "Master volume", STYPE_Dimmer, sSwitchGeneralSwitch, 8, NULL },
	{ "ZVL", "ZMT", "Zone 2 volume", STYPE_Dimmer, sSwitchGeneralSwitch, 8, NULL },
	{ "VL3", "MT3", "Zone 3 volume", STYPE_Dimmer, sSwitchGeneralSwitch, 8, NULL },
	{ "VL4", "MT4", "Zone 4 volume", STYPE_Dimmer, sSwitchGeneralSwitch, 8, NULL },
	{ "PWR", NULL, "Master power", STYPE_OnOff, sSwitchGeneralSwitch, 5, NULL },
	{ "ZPW", NULL, "Zone 2 power", STYPE_OnOff, sSwitchGeneralSwitch, 5, NULL },
	{ "PW3", NULL, "Zone 3 power", STYPE_OnOff, sSwitchGeneralSwitch, 5, NULL },
	{ "PW4", NULL, "Zone 4 power", STYPE_OnOff, sSwitchGeneralSwitch, 5, NULL },
	{ "SLI", NULL, "Master selector", STYPE_Selector, sSwitchTypeSelector, 5, NULL },
	{ "SLZ", NULL, "Zone 2 selector", STYPE_Selector, sSwitchTypeSelector, 5, NULL },
	{ "SL2", NULL, "Zone 3 selector", STYPE_Selector, sSwitchTypeSelector, 5, NULL },
	{ "SL3", NULL, "Zone 4 selector", STYPE_Selector, sSwitchTypeSelector, 5, NULL },
	{ "HDO", NULL, "HDMI Output", STYPE_Selector, sSwitchTypeSelector, 5, "SelectorStyle:0;LevelNames:Off|Main|Sub|Main+Sub;LevelOffHidden:true;LevelActions:00|01|02|03" },
```

Comment (gizmocuz) [reply hardware/OnkyoAVTCP.cpp:359]: For now please keep using your direct calls. I think there will be a function to send/update selector switches shortly, maybe then we can replace these functions

**Comment by gizmocuz** [comment]
Thanks!
